### DISCUSSION TITLE
untyped vaarg with macro

### DISCUSSION THREAD

**Discussion by LukyGuyLucky** [c3lang/c3c#1700] [Q&A] [answered]
With this example [https://github.com/c3lang/c3c/blob/master/test/test_suite/macros/macro_untyped_varargs_2.c3t](url)  I have two questions to ask for help:
1. Please refer to the running reslut.The order of the result of @foo(ping(1）,ping(-2),...) is not as same as argments pass to the macro order.In the handbook it says the evaluation order is  well-defined.
 ```Call arguments are evaluated in parameter order.```
So is this all right?
2.I am sorry ,but I just can't come out why there 2 more Ping[3141] in the printing result? 

```
Launching macro_untyped_varargs_2.exe
Ping[-1]    //<-------------Q1
Ping[1]    //<--------------
Ping[3141]
Ping[1000]
Ping[3141] //<----------Q2
Ping[3141] //<-----------
1
-1
3141
1000
int
double
105
33, 123
123, 33
Program completed with exit code 0.
```

**Comment by lerno** [comment, marked as answer]
That would normally be true, but macro vaargs are special. In a way they are all lazily evaluated. This is why they are following access order. Secondly, for anything *not* `$vaexpr` evaluation will only happen one time, then the value is "locked" in the call itself. So the `@foo(ping(1), ping(-1), ping(3141), ping(999 + 1));`

Makes the macro do the following evaluation:
1. Evaluate arg 1, since it appears first as a $vaarg (prints ping[-1])
2. Note that arg2 will be later evaluated as an $vaexpr
3. Evaluate arg 0, arg 2, arg 3 (prints ping[1], ping[3141] and ping[1000])
4. Enter macro
5. Use arg1 (already evaluated)
6. Evaluate the arg2 expression two times (this prints pint[3141] twice)
7. Print the values of the evaluation in step 1 and 3 in argument order (so 0 first, then 1 etc)

**Reply by LukyGuyLucky** [reply]
I got it now.Thank you!
